endpoints/v2/blob.py

```diff
@@ -51,82 +53,116 @@
-@parse_repository_name()
-@process_registry_jwt_auth(scopes=["pull"])
-@require_repo_read
-@anon_allowed
-@cache_control(max_age=31436000)
-def check_blob_exists(namespace_name, repo_name, digest):
-    # Find the blob.
-    blob = registry_model.get_cached_repo_blob(model_cache, namespace_name, repo_name, digest)
-    if blob is None:
-        raise BlobUnknown()
-
-    # Build the response headers.
-    headers = {
-        "Docker-Content-Digest": digest,
-        "Content-Length": blob.compressed_size,
-        "Content-Type": BLOB_CONTENT_TYPE,
-    }
-
-    # If our storage supports range requests, let the client know.
-    if storage.get_supports_resumable_downloads(blob.placements):
-        headers["Accept-Ranges"] = "bytes"
-
-    # Write the response to the client.
-    return Response(headers=headers)
-
-
-@v2_bp.route(BLOB_DIGEST_ROUTE, methods=["GET"])
-@disallow_for_account_recovery_mode
-@parse_repository_name()
-@process_registry_jwt_auth(scopes=["pull"])
-@require_repo_read
-@anon_allowed
-@check_region_blacklisted(BlobDownloadGeoBlocked)
-@cache_control(max_age=31536000)
-def download_blob(namespace_name, repo_name, digest):
-    # Find the blob.
-    blob = registry_model.get_cached_repo_blob(model_cache, namespace_name, repo_name, digest)
-    if blob is None:
-        raise BlobUnknown()
-
-    # Build the response headers.
-    headers = {"Docker-Content-Digest": digest}
-
-    # If our storage supports range requests, let the client know.
-    if storage.get_supports_resumable_downloads(blob.placements):
-        headers["Accept-Ranges"] = "bytes"
-
-    image_pulled_bytes.labels("v2").inc(blob.compressed_size)
-
-    # Short-circuit by redirecting if the storage supports it.
-    path = blob.storage_path
-    logger.debug("Looking up the direct download URL for path: %s", path)
-    direct_download_url = storage.get_direct_download_url(blob.placements, path, get_request_ip())
-    if direct_download_url:
-        logger.debug("Returning direct download URL")
-        resp = redirect(direct_download_url)
-        resp.headers.extend(headers)
-        return resp
-
-    # Close the database connection before we stream the download.
-    logger.debug("Closing database connection before streaming layer data")
-    headers.update(
-        {
+class CheckBlobExists(MethodView):
+    decorators = [
+        cache_control(max_age=31436000),
+        anon_allowed,
+        require_repo_read,
+        process_registry_jwt_auth(scopes=["pull"]),
+        parse_repository_name(),
+        disallow_for_account_recovery_mode,
+    ]
+
+    def head(self, namespace_name, repo_name, digest):
+        blob = registry_model.get_cached_repo_blob(model_cache, namespace_name, repo_name, digest)
+        if blob is None:
+            try:
+                proxy = setup_proxy(namespace_name, repo_name)
+                resp = proxy.blob_exists(digest)
+                return Response(**resp)
+            except ProxyNotSupported:
```

Comment:
Also, this block seems to be recurring in multiple endpoints. Could you make a decorator out of it to use on the relevant endpoints?

Reply:
Yeah, true.
This block will change a lot as we work on storing the images pulled from upstream registries. It's a bit early to know what's the best way to refactor it, so I rather we wait until the code takes more shape.